### DISCUSSION TITLE
Conditional Keyboard Shortcuts Configuration in Neovim

### DISCUSSION THREAD

**Discussion by jcastillopro** [NvChad/NvChad#2190] [Q&A]
Hello everyone,

I'm currently trying to leverage the flexibility of conditional keyboard shortcuts in Neovim, similar to the approach I've been using in Visual Studio Code with the `when` key.

In VSCode, I had the following configuration:

```json
{
    "key": "shift+t",
    "command": "workbench.action.focusActiveEditorGroup",
    "when": "explorerViewletVisible && filesExplorerFocus && !inputFocus"
}
```

This command provided a tailored behavior, allowing me to focus on the active editor group only when the explorer view is visible, files explorer has focus, and no input is currently in focus.

The core idea behind my inquiry is to be able to condition commands in a clean and clear way, thus altering the behavior of keys based on the scope or given condition. I'm curious to know if there's a parallel methodology or perhaps a plugin within Neovim that can offer this level of customization.

Any advice or guidance on how to replicate this behavior within the Neovim context would be greatly appreciated. 

Thanks in advance for your help.

**Comment by antoineco** [comment]
I believe the (Neo)Vim way to do this is to script those conditions and use them in the key mapping callback.

Here is an example of key mapping which includes conditionals: https://github.com/hrsh7th/nvim-cmp/blob/2743dd989e/lua/cmp/config/mapping.lua#L54-L63

**Reply by siduck** [reply]
@jcastillopro  elaborate, whats active editor group? and whats no input?

> This command provided a tailored behavior, allowing me to focus on the active editor group only when the explorer view is visible, files explorer has focus, and no input is currently in focus.

**Reply by siduck** [reply]
yo @antoineco thanks for actively helping nvchad users in issues/discussions here, saves my time!

**Reply by jcastillopro** [reply]
@antoineco @siduck  Thanks

It would be beneficial to have a feature akin to [when Clause Contexts](https://code.visualstudio.com/api/references/when-clause-contexts) available. Upon some investigation and using chatgpt, I've found an alternative that might work. However, I am not entirely sure if it is the correct or most efficient way to implement it. I would appreciate any guidance or suggestions for improvement. Typically, in my VSCode configuration, I use the same key for different purposes, depending on the active view.

```lua
-- In your conditional functions file (conditions.lua)

local M = {}

function M.can_execute_command()
  -- Here you put your logic to determine if the command can be executed
  -- Returns true if the command can be executed, false otherwise
end

return M
```

```lua
-- In your shortcuts file (shortcuts.lua)

-- Import your conditions module
local conditions = require('conditions')

local M = {}

M.general = {
  n = {
    -- your existing shortcuts...

    -- Example of how to use a function to check a condition before executing a command
    ["tu"] = { function() if conditions.can_execute_command() then vim.cmd('tabe') end end, "Create a new tab" },

    -- other shortcuts...
  },
  
  -- other modes...
}

return M
```

This way, you can add a 'when' clause context to neovim by defining a function in a separate `conditions.lua` file that returns `true` or `false` depending on whether a command can be executed under certain conditions. Then, in your `shortcuts.lua` file, you can check this condition before executing each command.


Thanks for the support.
